templates/components/agents/typescript/financial_report/app/api/chat/workflow/factory.ts

```diff
@@ -0,0 +1,82 @@
+import { WorkflowEvent } from "@llamaindex/core/workflow";
+import { Message } from "ai";
+import { ChatMessage, ToolCallLLM } from "llamaindex";
+import { getAnnotations } from "../llamaindex/streaming/annotations";
+import { FinancialReportWorkflow } from "./finReport";
+import { getAvailableTools } from "./tools";
+
+const TIMEOUT = 360 * 1000;
+
+class ResearchEvent extends WorkflowEvent<{ input: string }> {}
+class AnalyzeEvent extends WorkflowEvent<{ input: string }> {}
+class ReportEvent extends WorkflowEvent<{ input: string }> {}
+
+const prepareChatHistory = (chatHistory: Message[]): ChatMessage[] => {
+  // By default, the chat history only contains the assistant and user messages
+  // all the agents messages are stored in annotation data which is not visible to the LLM
+
+  const MAX_AGENT_MESSAGES = 10;
+  const agentAnnotations = getAnnotations<{ agent: string; text: string }>(
+    chatHistory,
+    { role: "assistant", type: "agent" },
+  ).slice(-MAX_AGENT_MESSAGES);
+
+  const agentMessages = agentAnnotations
+    .map(
+      (annotation) =>
+        `\n<${annotation.data.agent}>\n${annotation.data.text}\n</${annotation.data.agent}>`,
+    )
+    .join("\n");
+
+  const agentContent = agentMessages
+    ? "Here is the previous conversation of agents:\n" + agentMessages
+    : "";
+
+  if (agentContent) {
+    const agentMessage: ChatMessage = {
+      role: "assistant",
+      content: agentContent,
+    };
+    return [
+      ...chatHistory.slice(0, -1),
+      agentMessage,
+      chatHistory.slice(-1)[0],
+    ] as ChatMessage[];
+  }
+  return chatHistory as ChatMessage[];
+};
+
+export async function createWorkflow(options: {
+  chatHistory: Message[];
+  llm?: ToolCallLLM;
+  writeEvents?: boolean;
```

Comment:
`writeEvents` seems to be generally unused - is the idea a flag to define whether we emit `AgentRunEvens` or not? i think we don't need